### ISSUE TITLE
rule check bug in iptables versions used in the node-cache image causes growing iptables rules and high node load

### ISSUE THREAD

**Issue by elchenberg** [kubernetes/dns#568]
Summary: The rule check with iptables versions below 1.8.8 seems to be flaky, causing node-cache to add the nodelocaldns rules over and over again. The growing number of iptables rules causes a high load on the nodes. Updating iptables in the node-cache image to 1.8.9 seems to solve the issue.

---

We are using node-local-dns in our bare-metal Kubernetes clusters. We noticed the issue when we got a high load on our cluster nodes due to ksoftirqd due to growing iptables rules. Most of the rules (thousands) were duplicates of the nodelocaldns rules.

<details>
<summary>Debug log of our node-local-dns pods looked like this (this is when all of the rules already exist):</summary>

```
2023-02-01T15:30:45.825087775Z [DEBUG] iptables rule {raw PREROUTING [-p tcp -d 169.254.20.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:45.826258245Z [DEBUG] iptables rule {raw PREROUTING [-p udp -d 169.254.20.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:45.927684416Z [INFO] Added back nodelocaldns rule - {filter INPUT [-p tcp -d 169.254.20.10 --dport 53 -j ACCEPT]}
2023-02-01T15:30:46.025903438Z [INFO] Added back nodelocaldns rule - {filter INPUT [-p udp -d 169.254.20.10 --dport 53 -j ACCEPT]}
2023-02-01T15:30:46.028011661Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -s 169.254.20.10 --sport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:46.029096991Z [DEBUG] iptables rule {raw OUTPUT [-p udp -s 169.254.20.10 --sport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:46.164281006Z [INFO] Added back nodelocaldns rule - {filter OUTPUT [-p tcp -s 169.254.20.10 --sport 53 -j ACCEPT]}
2023-02-01T15:30:46.265490308Z [INFO] Added back nodelocaldns rule - {filter OUTPUT [-p udp -s 169.254.20.10 --sport 53 -j ACCEPT]}
2023-02-01T15:30:46.266599614Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -d 169.254.20.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:46.267605995Z [DEBUG] iptables rule {raw OUTPUT [-p udp -d 169.254.20.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:46.268615102Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -d 169.254.20.10 --dport 8080 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:46.269572682Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -s 169.254.20.10 --sport 8080 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:46.270554538Z [DEBUG] iptables rule {raw PREROUTING [-p tcp -d 10.96.0.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:46.272712023Z [DEBUG] iptables rule {raw PREROUTING [-p udp -d 10.96.0.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:46.369931607Z [INFO] Added back nodelocaldns rule - {filter INPUT [-p tcp -d 10.96.0.10 --dport 53 -j ACCEPT]}
2023-02-01T15:30:46.466682109Z [INFO] Added back nodelocaldns rule - {filter INPUT [-p udp -d 10.96.0.10 --dport 53 -j ACCEPT]}
2023-02-01T15:30:46.468432429Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -s 10.96.0.10 --sport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:46.470698258Z [DEBUG] iptables rule {raw OUTPUT [-p udp -s 10.96.0.10 --sport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:46.569405910Z [INFO] Added back nodelocaldns rule - {filter OUTPUT [-p tcp -s 10.96.0.10 --sport 53 -j ACCEPT]}
2023-02-01T15:30:46.673160237Z [INFO] Added back nodelocaldns rule - {filter OUTPUT [-p udp -s 10.96.0.10 --sport 53 -j ACCEPT]}
2023-02-01T15:30:46.674254844Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -d 10.96.0.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:46.686615108Z [DEBUG] iptables rule {raw OUTPUT [-p udp -d 10.96.0.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:46.687884514Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -d 10.96.0.10 --dport 8080 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:30:46.688804203Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -s 10.96.0.10 --sport 8080 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:45.824823003Z [DEBUG] iptables rule {raw PREROUTING [-p tcp -d 169.254.20.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:45.827053335Z [DEBUG] iptables rule {raw PREROUTING [-p udp -d 169.254.20.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:45.923006228Z [INFO] Added back nodelocaldns rule - {filter INPUT [-p tcp -d 169.254.20.10 --dport 53 -j ACCEPT]}
2023-02-01T15:31:46.027998462Z [INFO] Added back nodelocaldns rule - {filter INPUT [-p udp -d 169.254.20.10 --dport 53 -j ACCEPT]}
2023-02-01T15:31:46.029737982Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -s 169.254.20.10 --sport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:46.031194860Z [DEBUG] iptables rule {raw OUTPUT [-p udp -s 169.254.20.10 --sport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:46.13108496ing4Z [INFO] Added back nodelocaldns rule - {filter OUTPUT [-p tcp -s 169.254.20.10 --sport 53 -j ACCEPT]}
2023-02-01T15:31:46.230465108Z [INFO] Added back nodelocaldns rule - {filter OUTPUT [-p udp -s 169.254.20.10 --sport 53 -j ACCEPT]}
2023-02-01T15:31:46.231835424Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -d 169.254.20.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:46.233297673Z [DEBUG] iptables rule {raw OUTPUT [-p udp -d 169.254.20.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:46.234190792Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -d 169.254.20.10 --dport 8080 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:46.235199488Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -s 169.254.20.10 --sport 8080 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:46.236131930Z [DEBUG] iptables rule {raw PREROUTING [-p tcp -d 10.96.0.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:46.237110520Z [DEBUG] iptables rule {raw PREROUTING [-p udp -d 10.96.0.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:46.333346625Z [INFO] Added back nodelocaldns rule - {filter INPUT [-p tcp -d 10.96.0.10 --dport 53 -j ACCEPT]}
2023-02-01T15:31:46.429870611Z [INFO] Added back nodelocaldns rule - {filter INPUT [-p udp -d 10.96.0.10 --dport 53 -j ACCEPT]}
2023-02-01T15:31:46.432212894Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -s 10.96.0.10 --sport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:46.433356714Z [DEBUG] iptables rule {raw OUTPUT [-p udp -s 10.96.0.10 --sport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:46.529933800Z [INFO] Added back nodelocaldns rule - {filter OUTPUT [-p tcp -s 10.96.0.10 --sport 53 -j ACCEPT]}
2023-02-01T15:31:46.627303726Z [INFO] Added back nodelocaldns rule - {filter OUTPUT [-p udp -s 10.96.0.10 --sport 53 -j ACCEPT]}
2023-02-01T15:31:46.629235056Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -d 10.96.0.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:46.630709237Z [DEBUG] iptables rule {raw OUTPUT [-p udp -d 10.96.0.10 --dport 53 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:46.631639676Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -d 10.96.0.10 --dport 8080 -j NOTRACK]} for nodelocaldns already exists
2023-02-01T15:31:46.632524660Z [DEBUG] iptables rule {raw OUTPUT [-p tcp -s 10.96.0.10 --sport 8080 -j NOTRACK]} for nodelocaldns already exists
```
</details>

Our cluster nodes have the same version of iptables installed (`iptables v1.8.7 (nf_tables)`) and running the check in a loop showed that there was some flakiness:

```
while true; do date -Iseconds | tr '\n' ' '; iptables -v -C INPUT -t filter -p tcp -d 10.96.0.10 --dport 53 -j ACCEPT; sleep 1; done

2023-02-01T15:37:40+00:00 iptables: Bad rule (does a matching rule exist in that chain?).
2023-02-01T15:37:41+00:00 iptables: Bad rule (does a matching rule exist in that chain?).
2023-02-01T15:37:42+00:00 iptables: Bad rule (does a matching rule exist in that chain?).
2023-02-01T15:37:43+00:00 iptables: Bad rule (does a matching rule exist in that chain?).
2023-02-01T15:37:44+00:00 iptables: Bad rule (does a matching rule exist in that chain?).
2023-02-01T15:37:45+00:00 iptables: Bad rule (does a matching rule exist in that chain?).
2023-02-01T15:37:46+00:00 iptables: Bad rule (does a matching rule exist in that chain?).
2023-02-01T15:37:48+00:00 iptables: Bad rule (does a matching rule exist in that chain?).
2023-02-01T15:37:49+00:00 ACCEPT  tcp opt -- in * out *  0.0.0.0/0  -> 10.96.0.10   tcp dpt:53
2023-02-01T15:37:50+00:00 ACCEPT  tcp opt -- in * out *  0.0.0.0/0  -> 10.96.0.10   tcp dpt:53
2023-02-01T15:37:51+00:00 iptables: Bad rule (does a matching rule exist in that chain?).
```

When we updated the iptables version on the cluster nodes to 1.8.8 we were not able to reproduce the issue anymore on the hosts, but in the node-local-dns containers with version 1.8.7 the problem still persisted. It disappeared when we used our own node-cache image based on debian testing with iptables v1.8.9:

```Dockerfile
FROM debian:testing-slim
RUN set -eux; \
    apt-get update; \
    DEBIAN_FRONTEND=noninteractive apt-get install --assume-yes --no-install-recommends iptables; \
    apt-get clean; \
    rm -rf /var/lib/apt/lists/*
COPY --from=registry.k8s.io/dns/k8s-dns-node-cache:1.22.13 /node-cache /node-cache
EXPOSE 53 53/udp
EXPOSE 53 53/tcp
ENTRYPOINT ["/node-cache"]
```

Can someone reproduce the issue and would it be possbile to update iptables in the node-cache image?

- kubernetes version: v1.24.10
- node-cache image: `k8s.gcr.io/dns/k8s-dns-node-cache:1.22.17@sha256:c7f380975c95dc9abb61923fd4a923f89188e90838873ff412c2f5596b5ad15e`
- iptables version in the image: v1.8.7 (nf_tables)

**Comment by dpasiukevich** [comment]
We use `gcr.io/gke-release/distroless-iptables` which provides `iptables` binary. 

Could you ask in [kuberenetes/release](https://github.com/kubernetes/release/blob/master/images/build/distroless-iptables/OWNERS) repo, if they could make an image with iptables 1.8.8+?

I will then update the base image for nodelocaldns (or approve the PR if it's there).

But beforehand, there may be concerns with the iptables 1.8.8: https://twitter.com/n_copa/status/1570430510253608961

**Comment by skirsten** [comment]
Hi, this seems to be a issue again 😢.

I was experiencing the same problems with growing iptables rules and high node load on Ubuntu 24.04 nodes. The 22.04 nodes were fine. So I upgraded from `registry.k8s.io/dns/k8s-dns-node-cache:1.22.23` to `registry.k8s.io/dns/k8s-dns-node-cache:1.26.4` which also upgraded the iptables from `v1.8.7` to `v1.8.9`.

This fixed the issue on the 24.04 nodes, but now the 22.04 are broken and are growing the iptables rules.

Anybody know what is going on here and how to fix it?
Currently I am manually removing all iptable rules every few hours :(
```
sudo iptables-save | grep -v "NodeLocal DNS Cache" | sudo iptables-restore'
```
